src/ray/core_worker/core_worker.cc

```diff
@@ -248,6 +259,19 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
   }
 }
 
+void CoreWorker::HandleWorkerFailed(const WorkerID &worker_id,
+                                    const gcs::WorkerFailureData &worker_failed_data) {
+  RAY_LOG(DEBUG) << "Worker " << worker_id << " failed";
+  // TODO: Clean up after the failure:
+  // - If the failed worker is our owner, then exit.
+  // - If the failed worker is the owner for any active ObjectID references,
+  // then throw an UnreconstructableError for these objects. This can happen if
+  // the ObjectID was serialized.
```

Comment:
This is a performance optimization only right? Maybe we can skip it for simplicity.

Reply:
Right now it's not necessary because we timeout fetches that take too long. If we want to implement reconstruction for plasma objects, we'll have to change ray.get to either an infinite timeout (in which case we'll need to implement this feature), or poll the object's owner periodically to make sure the object is still being managed.

Note that all of that assumes ref counting for serialized object IDs, which also is not implemented currently.

Reply:
What I mean is doesn't the GetObjectStatus RPC already fail if the owner dies? So it's just a matter of whether the RPC failure callback fires first or this worker death one. The former is guaranteed to happen by gRPC, the latter is best effort and we can't rely on for correctness (which is what I meant by it's at best an optimization).

Reply:
@ericl  Would you mind expand a bit more on why the worker death notification is best-effort?

Reply:
@zhijunfu two reasons: it doesn't handle node failures, and also intentional exits, whereas checking via gRPC detects both.

Reply:
@ericl Thanks. Yes the current PR doesn't handle these. 
I was thinking we can handle node failures by adding a worker table to track which workers are on which nodes, like what actor table does currently.  For intentional exits, local raylet is aware of it and should be able handle that properly, while with grpc we are not able to know if it's intentional?

Reply:
Hmm, I think we could have this work without that by including the unique node id as part of the worker identifier. That way, broadcasting the node death is enough to identify any workers also killed.